### PR TITLE
Lexical scoping in checker

### PR DESCRIPTION
**Pull request by jamesr66a**
To facilitate ops with nested graph attributes, such as in https://github.com/onnx/onnx/pull/436

### CODE REVIEW

File: onnx/checker.cc [hunk 1]
```diff
@@ -236,6 +247,7 @@ void check_graph(const GraphProto& graph, const CheckerContext& ctx) {
   }
 
   std::unordered_set<std::string> output_names{};
+  // Inherit values avaiailable in outer scope
```
Comment (bddppq) [inline onnx/checker.cc]: nit: add a comment saying no shadowing

File: onnx/checker.h [hunk 2]
```diff
@@ -43,12 +49,25 @@ class CheckerContext {
   explicit CheckerContext() : ir_version(-1) {}
 };
 
+struct LexicalScopeContext {
+  std::unordered_set<std::string> output_names;
```
Comment (bddppq) [inline onnx/checker.h]: nit: maybe just "names"?

File: onnx/cpp2py_export.cc [hunk 3]
```diff
@@ -109,71 +110,80 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
...
       });
 
   checker.def(
-      "check_attribute", [](const py::bytes& bytes, const checker::CheckerContext& ctx) -> void {
+      "check_attribute",
+      [](const py::bytes& bytes, const checker::CheckerContext& ctx) -> void {
         std::unique_ptr<AttributeProto> proto(new AttributeProto());
         ParseProtoFromPyBytes(proto.get(), bytes);
-        checker::check_attribute(*proto, ctx);
+        checker::LexicalScopeContext lex_ctx;
```
Comment (bddppq) [inline onnx/cpp2py_export.cc]: nit: you can make it const or just create an anonymous one in the function call next line.

File: onnx/checker.h (hunk 2 above)
Comment (jamesr66a) [reply onnx/checker.h]: I named this output_names to be consistent with the code in the checker. maybe we should change both?

**Pull request merged**